cookbooks/ruby/libraries/helpers.rb

```diff
@@ -0,0 +1,176 @@
+    ruby_mask = ruby2x_mask(ruby_version)
+    use_mask_clear "ruby" do
+      mask_file "ruby"
+    end
+    use_mask ruby_mask do
+      mask_file "ruby"
+      only_if { ruby_mask }
+    end
+  end
+
+  def install_ruby_and_deps(reset_and_retry=true)
+    if need_ruby2x_reset
+      Chef::Log.info "reset the current ruby installation for a clean downgrade"
+      remove_nondefault_ruby_packages
+      reset_ruby_mask
+    end
+    component = node.engineyard.environment.ruby
+    ruby_dependencies = node.default[:ruby_dependencies]
+    packages = ruby_dependencies.merge({
+      component[:package] => component[:version]
+    })
+    package_atoms = packages.map { |package_name, package_version| "=#{package_name}-#{package_version}" }
+    execute 'install ruby and its dependencies' do
+      command %Q{emerge --read-news=n -g -n --color n --nospinner --quiet #{package_atoms.join(' ')}}
+      action :run
+    end
+  end
+
+  def uninstall_rubygems_update_gems(keep_version)
+    ruby_block "uninstall previous versions of rubygems" do
+      block do
+        output = Mixlib::ShellOut.new('gem list rubygems').run_command.stdout
+        matchdata = output.match(/\(([^\)]*)\)/)
+        if versions = (matchdata && matchdata[1])
+          versions.split(',').each do |version|
+            version.strip!
+            if keep_version.nil? || (version != keep_version)
+              Mixlib::ShellOut.new("gem uninstall rubygems-update -v '#{version}'").run_command
+            end
+          end
+        end
+      end
+    end
+  end
+
+  def reinstall_rubygems_os_package
+    ruby_dependencies = node.default[:ruby_dependencies]
+    rubygems_package = 'dev-ruby/rubygems'
+    rubygems_version = ruby_dependencies.fetch(rubygems_package, nil)
+    if not rubygems_version.nil?
+      package_atom = "=#{rubygems_package}-#{rubygems_version}"
+      execute 're-install rubygems' do
+        # NB: we call emerge without the -n (noreplace) flag here.
+        # It would prevent re-installing a package, which is the very thing we want to achieve here.
+        command %Q{emerge --read-news=n -g --color n --nospinner --quiet #{package_atom}}
+        action :run
+      end
+    end
+  end
+
+  def ensure_rubygems_version
+    rubygems = nil
+    rubygems = node.engineyard.environment.ruby.fetch(:rubygems, nil) if node.engineyard.environment.ruby?
+    # Specific rubygems version must be set in the ruby component, see cookbooks/ey-core/dnapi.rb.
+
+    if rubygems && (Gem::Version.new(`gem -v`) != Gem::Version.new(rubygems))
+      # Uninstall previously installed version of Rubygems
+      uninstall_rubygems_update_gems(rubygems)
+
+      ey_cloud_report "rubygems update" do
+        message "installing Rubygems #{rubygems}"
+      end
+
+      execute "install rubygems #{rubygems}" do
+        command "gem install rubygems-update -v #{rubygems}"
+      end
+
+      execute "update rubygems to >= #{rubygems}" do
+        command "update_rubygems"
```

Comment:
The old code that was replaced used `gem update --system #{rubygems}` (for rubygems greater than 1.5.2 which is always the case now on v5).

That said, calling update_rubygems *after* removing all other rubygems-update gems seems to do the same thing.

Reply:
Oh, I incorrectly parsed that cryptic if condition. @remotedevops should I change the command?

Reply:
We can leave it for now.